arangod/VocBase/vocbase.cpp

```diff
@@ -1871,6 +2040,55 @@ void TRI_SanitizeObjectWithEdges(VPackSlice const slice, VPackBuilder& builder)
   }
 }
 
+void TRI_vocbase_t::registerReplicatedLog(arangodb::replication2::LogId logId,
+                                          std::shared_ptr<arangodb::replication2::replicated_log::PersistedLog> persistedLog) {
+  std::unique_lock guard(_logManager->_mutex);
+  auto core = std::make_unique<arangodb::replication2::replicated_log::LogCore>(
+      std::move(persistedLog));
+  auto [iter, inserted] = _logManager->_logs.try_emplace(
+      logId, std::make_shared<replication2::replicated_log::ReplicatedLog>(
+                 std::move(core), server().getFeature<ReplicatedLogFeature>().metrics(),
+                 LoggerContext(Logger::REPLICATION2)));
+  server().getFeature<ReplicatedLogFeature>().metrics()->replicatedLogNumber->fetch_add(1);
+  TRI_ASSERT(inserted);
+}
+
+void TRI_vocbase_t::unregisterReplicatedLog(arangodb::replication2::LogId id) {
+  {
+    std::unique_lock guard(_logManager->_mutex);
+    _logManager->_logs.erase(id);
+  }
+  server().getFeature<ReplicatedLogFeature>().metrics()->replicatedLogNumber->fetch_sub(1);
+}
+
+auto TRI_vocbase_t::ensureReplicatedLog(arangodb::replication2::LogId id,
+                                        std::optional<std::string> const& collectionName)
+    -> std::shared_ptr<arangodb::replication2::replicated_log::ReplicatedLog> {
+  {
+    std::unique_lock guard(_logManager->_mutex);
+    if (auto iter = _logManager->_logs.find(id); iter != _logManager->_logs.end()) {
+      return iter->second;
+    }
+  }
+  // TODO vocbase is not thread safe for replicated logs
+  auto res = createReplicatedLog(id, collectionName);
```

Comment:
Isn't this prone to races? We first check under the mutex if a log with a certain id exists or not.
If we don't find it, we unlock and then create the replicated log without holding the mutex here.

Reply:
Fixed